### ISSUE TITLE
Question federated learning example

### ISSUE THREAD

**Issue by setsometso** [n1analytics/python-paillier#70]
Hi there,

i have a question on the federated learning example.

On this tutorial:
Szenario1:
https://blog.n1analytics.com/distributed-machine-learning-and-partially-homomorphic-encryption-1/
You get after running the federated learning function/protocol for each client a different MSE.

Szenario2:
On the current Code in the repository you get for every client after running the federated learning function/protocoll the same MSE value. 

Why is that and why should be that correct?

What is the difference between these two approaches?Could you explain that to me in natural language? Is Scenario1 false?Why should it be a problem to use federated learning on a pre-initialized model? In my case Scenario1 has a way better mse after federated learning on the local pre-initialized model

Do you have a classification example for that?

**Comment by wilko77** [comment]
Unfortunately, the blog post is somewhat inaccurate.
The problems lies in the initialization.

The way Gradient descent optimization works is:
 - pick some initial guess for a value of x (This can be random, or all zero, ...)
 - for some iterations do:
   - compute the gradient of your loss function at your current guess
   - It is know that the loss function decreases fastest in the direction of the negative gradient. That's why we now update our current guess by "moving" it in the direction of the negative gradient.
 
So how can this process be federated?
The main observation is that, the loss of a model θ over a dataset X is the sum of the losses of the model θ over each example in the dataset. Thus each data holder can compute its contribution to the gradient of the loss function for some model θ, and the sum of these contributions is the same as the loss of θ over the combined dataset.

Where did the blog post go wrong?
Aggregating the gradients only really makes sense if you compute them for the same model θ. We initialized the models on the clients with different values and thus the aggregated gradients were just gibberish. 

Why is scenario 2 better?
You can easily verify that in scenario 2 the model that each individual client learns is the same model as the one computed over the combined dataset. Thus there must have been a knowledge transfer between the clients.

> In my case Scenario1 has a way better mse after federated learning on the local pre-initialized model

The main reason why the MSEs went down is that the optimizer essentially ran for twice as many iterations in the federated case. If you train locally for the same number of iterations, then the MSEs will be very similar.

> Why should it be a problem to use federated learning on a pre-initialized model?
It is unclear if the aggregation of gradients of different models has a beneficial value for federated learning. You would have to carefully evaluate the utility of such an approach, as the results for common SGD won't apply. 
But that doesn't mean that this is impossible. In fact, there is a interesting idea called 'federated averaging' (Federated learning of deep networks using model averaging by McMahan etal.) where they propose to locally train models for a number of iterations and then periodically average the models between the parties. 

> Do you have a classification example for that?
robomotic suggested a logistic regression example in issue #63.
